### ISSUE TITLE
[Feature Proposal] Named Workspaces

### ISSUE THREAD

**Issue by LearnLinuxTV** [pop-os/cosmic-workspaces#29]
I was speaking to Levi about this and decided to put in a feature request to see if this is something that can be tackled. The basic idea here is to be able to re-name your workspaces, and maybe even assign certain apps to a specific workspace. This gives GNOME a more project-oriented workflow. On my end, I use one workspace per project.

For an example of this, see the following screenshot that I made:
![named_workspaces](https://user-images.githubusercontent.com/5324220/147609849-a654ac6c-945d-459f-a9b0-554800f81e4e.png)

That picture was made by me manually editing the folllowing gsettings key using dconf:

/org/gnome/desktop/wm/preferences/workspace-names

So, the functionality is already within GNOME, it's just not exposed anywhere obvious. Anyone can set that key to whatever they want. However, there are a few missing components to make this really work:

1.) The workspace identifier extension (built-in to GNOME) still only shows a number in the panel, instead of the workspace name. Part of this proposal would be to alter that to show the name in the panel instead.

2.) The workspace picker doesn't show the workspace label at all, which almost makes setting the key manually somewhat useless. Part of this proposal would be to display the workspace label underneath the workspace icon in the picker. Even better, allow the user to edit the names from the workspace picker too.

3.) Another change to make, is to set the minimum number of workspaces equal to however many workspaces someone has labeled. For example, if I named 10 workspaces, then anything after the tenth is dynamic. If I remove a couple of labels to where I only have 8 named workspaces, then the minimum number of dynamic workspaces changes to 8. This way, people can set names for workspaces but still benefit from dynamic workspaces beyond that.

4.) Extra bonus: Suggested by Levi, maybe have a way to assign specific apps to a specific workspace label. The first two items in this list would achieve the bulk of the feature, but having the ability to have an app open on a specific worksapce automatically would be even better.

In my experience, I find letting the user manually name their workspaces allows them to come up with their own project structure and easily sort their activities by project. The KDE Plasma desktop already does something like this, but I find it to be clunky, confusing, and tedious. Since GNOME already has a solid workspace implementation, simply adding the ability the remaining bits with the key that's already present would be a good productivity win IMO.

**Comment by jacobgkau** [comment]
I believe this has been on the roadmap for some time, although I don't see another issue open to track it yet.

**Comment by LearnLinuxTV** [comment]
> I believe this has been on the roadmap for some time, although I don't see another issue open to track it yet.

I seem to remember it being on the roadmap too, but I can't find where I read that. In any case, I think the fact that GNOME already has this feature half-implemented with the gsettings key I mentioned, should make this orders of magnitude easier to implement. I'm not sure if I'm right, but I assume that since there's a key you'd only need to expose it in the UI in the right places, and then you're done. I can do it myself if I had some 1 on 1 help.

**Comment by Mesum-Hussain** [comment]
I use the tiling window manager and wouldn't like this feature. I use a different Workspaces whenever one is full (with 4 apps open). Not necessarily use them as virtual desktops like in Windows or Mac OS.

**Comment by LearnLinuxTV** [comment]
> I use the tiling window manager and wouldn't like this feature. I use a different Workspaces whenever one is full (with 4 apps open). Not necessarily use them as virtual desktops like in Windows or Mac OS.

The beauty of this, is that if the feature doesn’t resonate with you, you can simply ignore it by not labeling your workspaces. In that case, there’s nothing different for you.

**Comment by Mesum-Hussain** [comment]
> > I use the tiling window manager and wouldn't like this feature. I use a different Workspaces whenever one is full (with 4 apps open). Not necessarily use them as virtual desktops like in Windows or Mac OS.
> 
> The beauty of this, is that if the feature doesn’t resonate with you, you can simply ignore it by not labeling your workspaces. In that case, there’s nothing different for you.

Agreed. And by the way I just noticed your profile name, are you the guy from the LearnLinuxTV YouTube channel?
I see your video, good work 👍🏻.

**Comment by LearnLinuxTV** [comment]
> Agreed. And by the way I just noticed your profile name, are you the guy from the LearnLinuxTV YouTube channel? I see your video, good work 👍🏻.

Thanks, I have a ton of fun 😀

**Comment by ChristopherDillon** [comment]
I would use this feature. Thanks for suggesting it!

**Comment by vnijs** [comment]
COSMIC looks amazing. I asked a question about "named workspaces" on [Reddit](https://www.reddit.com/r/pop_os/comments/1nrnf5k/cosmic_questions_tiling_and_workspace_labels/). This is available in hyprland and I use it for task/todo/project. Really nice feature (see screenshot below from the waybar). If, as @LearnLinuxTV mentioned, this feature is already available in GNOME it would be wonderful to see accessible in COSMIC as well. Thanks for making this amazing DE!

<img width="822" height="56" alt="Image" src="https://github.com/user-attachments/assets/4332356b-59c9-49c5-b3a9-aefba91788dc" />
